Source/DafnyLanguageServer/Workspace/Notifications/VerificationDiagnosticsParams.cs

```diff
@@ -0,0 +1,671 @@
+
+    public int GetNewChildrenCount() {
+      return NewChildren.Count;
+    }
+
+    public void AddNewChild(NodeDiagnostic newChild) {
+      NewChildren.Add(newChild);
+    }
+
+    public void SaveNewChildren() {
+      Children = NewChildren;
+      ResetNewChildren();
+    }
+
+    public void ResetNewChildren() {
+      NewChildren = new();
+    }
+
+    public NodeDiagnostic SetObsolete() {
+      if (StatusCurrent != CurrentStatus.Obsolete) {
+        StatusCurrent = CurrentStatus.Obsolete;
+        foreach (var child in Children) {
+          child.SetObsolete();
+        }
+      }
+
+      return this;
+    }
+
+    // Returns true if it started the method, false if it was already started
+    public virtual bool Start() {
+      if (StatusCurrent != CurrentStatus.Verifying || !Started) {
+        StartTime = DateTime.Now;
+        StatusCurrent = CurrentStatus.Verifying;
+        foreach (var child in Children) {
+          child.Start();
+        }
+        Started = true;
+        return true;
+      }
+
+      return false;
+    }
+
+    // Returns true if it did stop the current node, false if it was already stopped
+    public virtual bool Stop() {
+      if (StatusCurrent != CurrentStatus.Current || !Finished) {
+        EndTime = DateTime.Now;
+        StatusCurrent = CurrentStatus.Current;
+        foreach (var child in Children) {
+          child.Stop();
+        }
+        Finished = true;
+        return true;
+      }
+
+      return false;
+    }
+
+    public void PropagateChildrenErrorsUp() {
+      var childrenHaveErrors = false;
+      foreach (var child in Children) {
+        child.PropagateChildrenErrorsUp();
+        if (child.StatusVerification == VerificationStatus.Error) {
+          childrenHaveErrors = true;
+        }
+      }
+
+      if (childrenHaveErrors) {
+        StatusVerification = VerificationStatus.Error;
+      }
+    }
+
+    // Requires PropagateChildrenErrorsUp to have been called before.
+    public virtual void RenderInto(LineVerificationStatus[] perLineDiagnostics, bool contextHasErrors = false, bool contextIsPending = false, Range? otherRange = null, Range? contextRange = null) {
+      Range range = otherRange ?? Range;
+      var isSingleLine = range.Start.Line == range.End.Line;
+      LineVerificationStatus targetStatus = StatusVerification switch {
+        VerificationStatus.Unknown => StatusCurrent switch {
+          CurrentStatus.Current => LineVerificationStatus.Unknown,
+          CurrentStatus.Obsolete => LineVerificationStatus.Scheduled,
+          CurrentStatus.Verifying => LineVerificationStatus.Verifying,
+          _ => throw new ArgumentOutOfRangeException()
+        },
+        // let's be careful to no display "Verified" for a range if the context does not have errors and is pending
+        // because there might be other errors on the same range.
+        VerificationStatus.Verified => StatusCurrent switch {
```

Comment:
I have a feeling this logic could be simplified, perhaps by defining some relationships between status constants.